astester.imports.js

```diff
@@ -30,31 +66,94 @@ export function local(memory) {
     massa: {
       memory,
       assembly_script_generate_event(string) {
-        console.log("event", getString(string));
+        console.log('event', getString(string));
+      },
+
+      assembly_script_set_data(k_ptr, v_ptr) {
+        const k = getString(k_ptr);
+        const v = getString(v_ptr);
+        const addressStorage = ledger.get(defaultAddress).storage;
+        addressStorage.set(k, v);
       },
 
       assembly_script_set_data_for(a_ptr, k_ptr, v_ptr) {
         const a = getString(a_ptr);
         const k = getString(k_ptr);
         const v = getString(v_ptr);
-        if (!Storage.has(a)) {
-          Storage.set(a, new Map());
+        if (!ledger.has(a)) {
+          ledger.set(a, emptyAddress);
         }
-        const addressStorage = Storage.get(a);
+
+        const addressStorage = ledger.get(a).storage;
         addressStorage.set(k, v);
       },
+
+      assembly_script_get_data(k_ptr) {
+        let v = '';
+        const k = getString(k_ptr);
+        if (ledger.has(defaultAddress)) {
+          const addressStorage = ledger.get(defaultAddress).storage;
+          if (addressStorage.has(k)) {
+            v = addressStorage.get(k);
+          }
+        }
+        return newString(v);
+      },
+
       assembly_script_get_data_for(a_ptr, k_ptr) {
-        let v = "";
+        let v = '';
         const a = getString(a_ptr);
         const k = getString(k_ptr);
-        if (Storage.has(a)) {
-          const addressStorage = Storage.get(a);
+        if (ledger.has(a)) {
+          const addressStorage = ledger.get(a).storage;
           if (addressStorage.has(k)) {
             v = addressStorage.get(k);
           }
         }
         return newString(v);
       },
+
+      assembly_script_has_data(k_ptr) {
+        const k = getString(k_ptr);
+        const addressStorage = ledger.get(defaultAddress).storage;
+        return addressStorage.has(k);
+      },
+
+      assembly_script_has_data_for(a_ptr, k_ptr) {
+        const a = getString(a_ptr);
+        const k = getString(k_ptr);
+        if (ledger.has(a)) {
+          const addressStorage = ledger.get(a).storage;
+          return addressStorage.has(k);
+        } else {
+          return false;
+        }
+      },
+
+      assembly_script_get_call_stack() {
+        const [_, contractAddress] = ledger.keys();
```

Comment:
This raise two questions:
- are you sure this function is determinist ? If the ledger is an hashMap the answer is no.
- what if you add multiple address using the getOf function ?

My proposal is to set automatically the contractAddress, and let the caller overwrite it if needed (can be done in another issue).

Reply:
Yeah I agree, I'll modify it.
I'm currently facing issues because of this

Reply:
Changed